target/ppc/translate/vmx-impl.c.inc

```diff
@@ -1008,41 +1008,263 @@ static void glue(gen_, name0##_##name1)(DisasContext *ctx)             \
     }                                                                  \
 }
 
-GEN_VXRFORM(vcmpequb, 3, 0)
-GEN_VXRFORM(vcmpequh, 3, 1)
-GEN_VXRFORM(vcmpequw, 3, 2)
-GEN_VXRFORM(vcmpequd, 3, 3)
-GEN_VXRFORM(vcmpnezb, 3, 4)
-GEN_VXRFORM(vcmpnezh, 3, 5)
-GEN_VXRFORM(vcmpnezw, 3, 6)
-GEN_VXRFORM(vcmpgtsb, 3, 12)
-GEN_VXRFORM(vcmpgtsh, 3, 13)
-GEN_VXRFORM(vcmpgtsw, 3, 14)
-GEN_VXRFORM(vcmpgtsd, 3, 15)
-GEN_VXRFORM(vcmpgtub, 3, 8)
-GEN_VXRFORM(vcmpgtuh, 3, 9)
-GEN_VXRFORM(vcmpgtuw, 3, 10)
-GEN_VXRFORM(vcmpgtud, 3, 11)
+static void do_vcmp_rc(int vrt)
+{
+    TCGv_i64 t0, t1;
+
+    t0 = tcg_temp_new_i64();
+    t1 = tcg_temp_new_i64();
+
+    get_avr64(t0, vrt, true);
+    tcg_gen_ctpop_i64(t1, t0);
+    get_avr64(t0, vrt, false);
+    tcg_gen_ctpop_i64(t0, t0);
+    tcg_gen_add_i64(t1, t0, t1);
+
+    tcg_gen_setcondi_i64(TCG_COND_EQ, t0, t1, 0);
+    tcg_gen_shli_i64(t0, t0, 1);
+
+    tcg_gen_setcondi_i64(TCG_COND_EQ, t1, t1, 128);
+    tcg_gen_shli_i64(t1, t1, 3);
+
+    tcg_gen_or_i64(t0, t0, t1);
+    tcg_gen_extrl_i64_i32(cpu_crf[6], t0);
+
+    tcg_temp_free_i64(t0);
+    tcg_temp_free_i64(t1);
+}
+
+static bool do_vcmp(DisasContext *ctx, arg_VC *a, TCGCond cond, int vece)
+{
+    REQUIRE_INSNS_FLAGS(ctx, ALTIVEC);
+    REQUIRE_VECTOR(ctx);
+
+    tcg_gen_gvec_cmp(cond, vece, avr_full_offset(a->vrt),
+                     avr_full_offset(a->vra), avr_full_offset(a->vrb), 16, 16);
+    tcg_gen_gvec_shli(vece, avr_full_offset(a->vrt), avr_full_offset(a->vrt),
+                      (8 << vece) - 1, 16, 16);
+    tcg_gen_gvec_sari(vece, avr_full_offset(a->vrt), avr_full_offset(a->vrt),
+                      (8 << vece) - 1, 16, 16);
+
+    if (a->rc) {
+        do_vcmp_rc(a->vrt);
+    }
+
+    return true;
+}
+
+TRANS(VCMPEQUB, do_vcmp, TCG_COND_EQ, MO_8)
+TRANS(VCMPEQUH, do_vcmp, TCG_COND_EQ, MO_16)
+TRANS(VCMPEQUW, do_vcmp, TCG_COND_EQ, MO_32)
+TRANS(VCMPEQUD, do_vcmp, TCG_COND_EQ, MO_64)
+
+TRANS(VCMPGTSB, do_vcmp, TCG_COND_GT, MO_8)
+TRANS(VCMPGTSH, do_vcmp, TCG_COND_GT, MO_16)
+TRANS(VCMPGTSW, do_vcmp, TCG_COND_GT, MO_32)
+TRANS(VCMPGTSD, do_vcmp, TCG_COND_GT, MO_64)
+TRANS(VCMPGTUB, do_vcmp, TCG_COND_GTU, MO_8)
+TRANS(VCMPGTUH, do_vcmp, TCG_COND_GTU, MO_16)
+TRANS(VCMPGTUW, do_vcmp, TCG_COND_GTU, MO_32)
+TRANS(VCMPGTUD, do_vcmp, TCG_COND_GTU, MO_64)
+
+TRANS(VCMPNEB, do_vcmp, TCG_COND_NE, MO_8)
+TRANS(VCMPNEH, do_vcmp, TCG_COND_NE, MO_16)
+TRANS(VCMPNEW, do_vcmp, TCG_COND_NE, MO_32)
```

Comment:
Could you review if the `REQUIRE_INSNS_FLAGS` and `REQUIRE_INSNS_FLAGS2` of these instructions are correct? I think it might have been wrong before already. For example `VCMPNEW` was added in ISAv3.0 but `do_vcmp` is not checking for it. I don't know if this was intended or not.